File: src/main/java/org/cobbzilla/util/io/UniqueFileFsWalker.java
package org.cobbzilla.util.io;

import org.slf4j.Logger;

import java.io.File;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

import static org.cobbzilla.util.io.FileUtil.abs;
import static org.cobbzilla.util.security.ShaUtil.sha256_file;

public class UniqueFileFsWalker implements FilesystemVisitor {

    private static final Logger log = org.slf4j.LoggerFactory.getLogger(UniqueFileFsWalker.class);
    private Map<String, Set<String>> hash;

    public UniqueFileFsWalker (int size) { hash = new ConcurrentHashMap<>(size); }

    @Override public void visit(File file) {
        final String path = abs(file);
        log.debug(path);
        hash.computeIfAbsent(sha256_file(file), k -> new HashSet<>()).add(path);
    }

    public Map<String, Set<String>> getHash() {
        return this.hash;
    }
}

File: src/main/java/org/cobbzilla/util/io/UniqueFileFsWalker.kt
package org.cobbzilla.util.io

import org.slf4j.Logger

import java.io.File
import java.util.HashSet
import java.util.concurrent.ConcurrentHashMap

import org.cobbzilla.util.io.FileUtil.abs
import org.cobbzilla.util.security.ShaUtil.sha256_file

class UniqueFileFsWalker(size: Int) : FilesystemVisitor {
    private val hash: MutableMap<String, Set<String>>

    init {
        hash = ConcurrentHashMap(size)
    }

    override fun visit(file: File) {
        val path = abs(file)
        log.debug(path)
        (hash as java.util.Map<String, Set<String>>).computeIfAbsent(sha256_file(file)) { k -> HashSet() }.add(path)
    }

    fun getHash(): Map<String, Set<String>> {
        return this.hash
    }

    companion object {

        private val log = org.slf4j.LoggerFactory.getLogger(UniqueFileFsWalker::class.java)
    }
}
